pallets/pallet-bonded-coins/src/lib.rs

```diff
@@ -199,32 +214,201 @@ pub mod pallet {
+
+			ensure!(pool_details.state.is_destroying(), Error::<T>::InUse);
+
+			for currency_id in pool_details.bonded_currencies {
+				if T::Fungibles::asset_exists(currency_id.clone()) {
+					// This would fail with an InUse error if there are any accounts left on any currency
+					T::Fungibles::finish_destroy(currency_id.clone())?;
+				}
+			}
+
+			let pool_account = pool_id.clone().into();
+
+			let total_collateral_issuance = Self::get_pool_collateral(&pool_account);
+
+			if !total_collateral_issuance.is_zero() {
+				T::CollateralCurrency::transfer(
+					T::CollateralAssetId::get(),
+					&pool_account,
+					&pool_details.manager,
+					total_collateral_issuance,
+					Preservation::Expendable,
+				)?;
+			}
+
+			Pools::<T>::remove(&pool_id);
+
+			// TODO: refund deposit
+
+			Self::deposit_event(Event::Destroyed(pool_id));
+
+			Ok(())
 		}
 	}
 
-	impl<T: Config> Pallet<T> {}
+	impl<T: Config> Pallet<T> {
+		fn do_start_destroy_pool(pool_id: &T::PoolId, maybe_check_manager: Option<&AccountIdOf<T>>) -> DispatchResult {
+			Pools::<T>::try_mutate(pool_id, |maybe_pool_details| -> DispatchResult {
+				let pool_details = maybe_pool_details.as_mut().ok_or(Error::<T>::PoolUnknown)?;
+
+				ensure!(!pool_details.state.is_destroying(), Error::<T>::Destroying);
+
+				if let Some(caller) = maybe_check_manager {
+					ensure!(pool_details.is_manager(caller), Error::<T>::Unauthorized);
+				}
+
+				pool_details.state.destroy();
+
+				Self::deposit_event(Event::DestructionStarted(pool_id.clone()));
+
+				let total_collateral_issuance = Self::get_pool_collateral(&pool_id.clone().into());
+
+				if total_collateral_issuance.is_zero() {
+					// no collateral to refund; deactivate all currencies and emit RefundCompleted
+					for currency_id in pool_details.bonded_currencies.iter() {
+						T::Fungibles::start_destroy(currency_id.clone(), None)?;
+					}
+
+					Self::deposit_event(Event::RefundCompleted(pool_id.clone()));
+
+					return Ok(());
+				}
+
+				// retrieve total supply of linked bonded currencies and deactivate those where supply is 0
+				let mut non_zero_supply = false;
+				for currency_id in pool_details.bonded_currencies.iter() {
+					if T::Fungibles::total_issuance(currency_id.clone()).is_zero() {
+						// supply is 0, currency can be deactivated
+						T::Fungibles::start_destroy(currency_id.clone(), None)?;
+					} else {
+						// if at least one currency has a non-zero supply, we must call refund_account next -> do not emit event
+						non_zero_supply = true;
+					}
+				}
+
+				// if all currencies have a 0 supply and have been deactivated, refund_account can be skipped -> emit RefundCompleted
+				if !non_zero_supply {
+					Self::deposit_event(Event::RefundCompleted(pool_id.clone()));
+				}
+
+				Ok(())
+			})
+		}
+
+		fn get_pool_collateral(pool_account: &AccountIdOf<T>) -> CollateralCurrencyBalanceOf<T> {
+			T::CollateralCurrency::total_balance(T::CollateralAssetId::get(), pool_account)
+		}
```

Comment:
Could we maybe make this function generic against the Currency? I would then also use it in the mint burn process for the bonded fungibles.

Reply:
what does that mean exactly? but actually we may not need this function at all if we switch to the `fungible` trait, bc then we don't have to look up an asset id

Reply:
it seems that with the `fungible` trait we may have an issue with `touch`ing though, as the `AssetTouch` trait requires an asset id.

Reply:
I think we will stick with the traits from the `fungibles` no? If we want to allow more than one collateral, we have to provide the corresponding `asset_id`